services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVncClient.java

```diff
@@ -121,22 +126,33 @@ public void run() {
                         s_logger.error("Unexpected exception", e);
                     }
 
-                    String ver = client.handshake();
-                    session.getRemote().sendBytes(ByteBuffer.wrap(ver.getBytes(), 0, ver.length()));
+                    String ver;
+                    byte[] b;
+                    if (!client.isVncOverWebSocketConnection()) {
+                        ver = client.handshake();
+                        session.getRemote().sendBytes(ByteBuffer.wrap(ver.getBytes(), 0, ver.length()));
 
-                    byte[] b = client.authenticate(getClientHostPassword());
-                    session.getRemote().sendBytes(ByteBuffer.wrap(b, 0, 4));
+                        b = client.authenticate(getClientHostPassword());
+                        session.getRemote().sendBytes(ByteBuffer.wrap(b, 0, 4));
+                    }
 
                     int readBytes;
                     while (connectionAlive) {
-                        b = new byte[100];
-                        readBytes = client.read(b);
-                        if (readBytes == -1) {
-                            break;
-                        }
-                        if (readBytes > 0) {
-                            session.getRemote().sendBytes(ByteBuffer.wrap(b, 0, readBytes));
-                            updateFrontEndActivityTime();
+                        if (client.isVncOverWebSocketConnection()) {
+                            if (client.isVncOverWebSocketConnectionOpen()) {
+                                updateFrontEndActivityTime();
+                            }
+                            connectionAlive = client.isVncOverWebSocketConnectionAlive();
+                        } else {
+                            b = new byte[100];
+                            readBytes = client.read(b);
+                            if (readBytes == -1) {
+                                break;
+                            }
+                            if (readBytes > 0) {
+                                session.getRemote().sendBytes(ByteBuffer.wrap(b, 0, readBytes));
+                                updateFrontEndActivityTime();
+                            }
```

Comment:
could you disect `initlient()` into multiple methods, it is sneakily growing too complex.

Reply:
Done, thanks